src/fps/proposal/MultisigProposal.sol

```diff
@@ -0,0 +1,854 @@
+            console.logBytes(dataToSign);
+        }
+    }
+
+    /// @notice print the hash to approve by EOA for nested multisig
+    function printNestedHashToApprove() public view {
+        bytes memory callData = _generateApproveMulticallData();
+
+        for (uint256 i; i < startingOwners.length; i++) {
+            bytes32 hash = keccak256(_getDataToSign(startingOwners[i], callData));
+            console.log("Nested multisig: %s", _getAddressLabel(startingOwners[i]));
+            console.logBytes32(hash);
+        }
+    }
+
+    /// --------------------------------------------------------------------
+    /// --------------------------------------------------------------------
+    /// ------------------------- Internal functions -----------------------
+    /// --------------------------------------------------------------------
+    /// --------------------------------------------------------------------
+
+    /// @notice validate actions inclusion
+    /// default implementation check for duplicate actions
+    function _validateAction(address target, uint256 value, bytes memory data) internal virtual {
+        uint256 actionsLength = actions.length;
+        for (uint256 i = 0; i < actionsLength; i++) {
+            // Check if the target, arguments and value matches with other existing actions.
+            bool isDuplicateTarget = actions[i].target == target;
+            bool isDuplicateArguments = keccak256(actions[i].arguments) == keccak256(data);
+            bool isDuplicateValue = actions[i].value == value;
+
+            require(!(isDuplicateTarget && isDuplicateArguments && isDuplicateValue), "Duplicated action found");
+        }
+    }
+
+    /// @notice print the calldata to be executed by safe
+    function _printProposalCalldata() internal virtual {
+        console.log("\n\n------------------ Proposal Calldata ------------------");
+        console.logBytes(getCalldata());
+    }
+
+    /// @notice helper function to generate the approveHash calldata to be executed by child multisig owner on parent multisig
+    function _generateApproveMulticallData() internal view returns (bytes memory) {
+        bytes32 hash = keccak256(_getDataToSign(multisig, getCalldata()));
+        Call3Value memory call = Call3Value({
+            target: multisig,
+            allowFailure: false,
+            value: 0,
+            callData: abi.encodeCall(IGnosisSafe(multisig).approveHash, (hash))
+        });
+
+        Call3Value[] memory calls = new Call3Value[](1);
+        calls[0] = call;
+        return abi.encodeWithSignature("aggregate3Value((address,bool,uint256,bytes)[])", calls);
+    }
+
+    /// @notice helper function to prepare the signatures to be executed
+    /// @param _safe The address of the parent multisig
+    /// @param hash The hash to be approved
+    /// @return The signatures to be executed
+    function prepareSignatures(address _safe, bytes32 hash) internal view returns (bytes memory) {
```

Comment:
Base has a Signatures.sol library for this that maybe we can reuse here to reduce the size of this file. https://github.com/base-org/contracts/blob/main/script/universal/Signatures.sol

Reply:
It looks like this code is copy/pasted from the base repo, so we may as well just continue using the base repo as a dep to make this PR easier to read and this file easier to maintain. We don't need to move away from deps just to reduce the number of dependencies, we only need to move away from deps if doing so solves a problem for us

Reply:
Nice, I think it would be nice to keep this file as minimal as possible. Being able to reason about the logic in here easily will pay dividends down the line as we want to add more features.

Reply:
This is completed so resolving this comment.